### PR TITLE
Bug 1810047 - Fixup WebExtensionPromptFeature

### PR DESCRIPTION
**Pull request by willdurand**
Sorry this is something I should have done in https://github.com/mozilla-mobile/firefox-android/pull/3917. Not sure what I was thinking when I landed the PR as is.

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1810047

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt [hunk 1]
```diff
@@ -302,6 +314,20 @@ class WebExtensionPromptFeature(
...
+        consumePromptRequest()
+    }
+
+    private fun handleDeniedOptionalPermissions(
+        promptRequest: WebExtensionPromptRequest.AfterInstallation.OptionalPermissions,
+    ) {
+        promptRequest.onConfirm(false)
+        consumePromptRequest()
+    }
+
```
Comment (Amejia481) [inline fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt:330]: Noticed we had some duplication between the permissions handling, as they don't belong to the same object hierarchy, it was a bit cumbersome to handle. Below I added a diff, where I put them into the same hierarchy reducing the duplicated code, and make it a bit clear to work with them. What do you think?   


```diff
diff --git a/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt b/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt
index 87152b3c42..c22c2cab3f 100644
--- a/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt
+++ b/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt
@@ -38,27 +38,34 @@ sealed class WebExtensionPromptRequest {
      * @param extension The [WebExtension] that requested the dialog to be shown.
      */
     sealed class AfterInstallation(open val extension: WebExtension) : WebExtensionPromptRequest() {
-        /**
-         * Value type that represents a request for a permission prompt.
-         * @property extension The [WebExtension] that requested the dialog to be shown.
-         * @property onConfirm A callback indicating whether the permissions were granted or not.
-         */
-        data class Permissions(
-            override val extension: WebExtension,
-            val onConfirm: (Boolean) -> Unit,
-        ) : AfterInstallation(extension)
 
-        /**
-         * Value type that represents a request for a (optional) permission prompt.
-         * @property extension The [WebExtension] that requested the dialog to be shown.
-         * @property permissions The optional permissions to list in the dialog.
-         * @property onConfirm A callback indicating whether the permissions were granted or not.
-         */
-        data class OptionalPermissions(
+
+        sealed class Permissions(
             override val extension: WebExtension,
-            val permissions: List<String>,
-            val onConfirm: (Boolean) -> Unit,
-        ) : AfterInstallation(extension)
+            open val onConfirm: (Boolean) -> Unit,
+        ) : AfterInstallation(extension) {
+            /**
+             * Value type that represents a request for a permission prompt.
+             * @property extension The [WebExtension] that requested the dialog to be shown.
+             * @property onConfirm A callback indicating whether the permissions were granted or not.
+             */
+            data class Required(
+                override val extension: WebExtension,
+                override val onConfirm: (Boolean) -> Unit,
+            ) : Permissions(extension, onConfirm)
+
+            /**
+             * Value type that represents a request for a (optional) permission prompt.
+             * @property extension The [WebExtension] that requested the dialog to be shown.
+             * @property permissions The optional permissions to list in the dialog.
+             * @property onConfirm A callback indicating whether the permissions were granted or not.
+             */
+            data class Optional(
+                override val extension: WebExtension,
+                val permissions: List<String>,
+                override val onConfirm: (Boolean) -> Unit,
+            ) : Permissions(extension, onConfirm)
+        }
 
         /**
          * Value type that represents a request for showing post-installation prompt.
diff --git a/fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt b/fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt
index 674d5734d3..86f99ca929 100644
--- a/fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt
+++ b/fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt
@@ -77,11 +77,11 @@ class WebExtensionPromptFeature(
         // opens the add-ons manager.
         val addon = Addon.newFromWebExtension(promptRequest.extension)
         when (promptRequest) {
-            is WebExtensionPromptRequest.AfterInstallation.Permissions -> handlePermissionRequest(
+            is WebExtensionPromptRequest.AfterInstallation.Permissions.Required -> handleRequiredPermissionRequest(
                 addon,
                 promptRequest,
             )
-            is WebExtensionPromptRequest.AfterInstallation.OptionalPermissions -> handleOptionalPermissionsRequest(
+            is WebExtensionPromptRequest.AfterInstallation.Permissions.Optional -> handleOptionalPermissionsRequest(
                 addon,
                 promptRequest,
             )
@@ -108,34 +108,34 @@ class WebExtensionPromptFeature(
         showPostInstallationDialog(addon)
     }
 
-    private fun handlePermissionRequest(
+    private fun handleRequiredPermissionRequest(
         addon: Addon,
         promptRequest: WebExtensionPromptRequest.AfterInstallation.Permissions,
     ) {
         showPermissionDialog(
             addon = addon,
-            onPositiveButtonClicked = { handleApprovedPermissions(promptRequest) },
-            onNegativeButtonClicked = { handleDeniedPermissions(promptRequest) },
+            onPositiveButtonClicked = { handlePermissions(promptRequest, granted = true) },
+            onNegativeButtonClicked = { handlePermissions(promptRequest, granted = true) },
         )
     }
 
     @VisibleForTesting
     internal fun handleOptionalPermissionsRequest(
         addon: Addon,
-        promptRequest: WebExtensionPromptRequest.AfterInstallation.OptionalPermissions,
+        promptRequest: WebExtensionPromptRequest.AfterInstallation.Permissions.Optional,
     ) {
         val shouldGrantWithoutPrompt = Addon.localizePermissions(promptRequest.permissions, context).isEmpty()
 
         // If we don't have any promptable permissions, just proceed.
         if (shouldGrantWithoutPrompt) {
-            handleApprovedOptionalPermissions(promptRequest)
+            handlePermissions(promptRequest, granted = true)
             return
         }
 
         showPermissionDialog(
             addon = addon,
-            onPositiveButtonClicked = { handleApprovedOptionalPermissions(promptRequest) },
-            onNegativeButtonClicked = { handleDeniedOptionalPermissions(promptRequest) },
+            onPositiveButtonClicked = { handlePermissions(promptRequest, granted = true) },
+            onNegativeButtonClicked = { handlePermissions(promptRequest, granted = true) },
             forOptionalPermissions = true,
             optionalPermissions = promptRequest.permissions,
         )
@@ -252,31 +252,15 @@ class WebExtensionPromptFeature(
         findPreviousPermissionDialogFragment()?.let { dialog ->
             dialog.onPositiveButtonClicked = { addon ->
                 store.state.webExtensionPromptRequest?.let { promptRequest ->
-                    when (promptRequest) {
-                        is WebExtensionPromptRequest.AfterInstallation.Permissions -> {
-                            if (addon.id == promptRequest.extension.id) {
-                                handleApprovedPermissions(promptRequest)
-                            }
-                        }
-                        is WebExtensionPromptRequest.AfterInstallation.OptionalPermissions -> {
-                            if (addon.id == promptRequest.extension.id) {
-                                handleApprovedOptionalPermissions(promptRequest)
-                            }
-                        }
-                        else -> { /* nothing to do for the other requests. */ }
+                    if (promptRequest is WebExtensionPromptRequest.AfterInstallation.Permissions && addon.id == promptRequest.extension.id) {
+                        handlePermissions(promptRequest, granted = true)
                     }
                 }
             }
             dialog.onNegativeButtonClicked = {
                 store.state.webExtensionPromptRequest?.let { promptRequest ->
-                    when (promptRequest) {
-                        is WebExtensionPromptRequest.AfterInstallation.Permissions -> {
-                            handleDeniedPermissions(promptRequest)
-                        }
-                        is WebExtensionPromptRequest.AfterInstallation.OptionalPermissions -> {
-                            handleDeniedOptionalPermissions(promptRequest)
-                        }
-                        else -> { /* nothing to do for the other requests. */ }
+                    if (promptRequest is WebExtensionPromptRequest.AfterInstallation.Permissions) {
+                        handlePermissions(promptRequest, granted = false)
                     }
                 }
             }
@@ -304,27 +288,9 @@ class WebExtensionPromptFeature(
         }
     }
 
-    private fun handleDeniedPermissions(promptRequest: WebExtensionPromptRequest.AfterInstallation.Permissions) {
-        promptRequest.onConfirm(false)
-        consumePromptRequest()
-    }
-
-    private fun handleApprovedPermissions(promptRequest: WebExtensionPromptRequest.AfterInstallation.Permissions) {
-        promptRequest.onConfirm(true)
-        consumePromptRequest()
-    }
 
-    private fun handleApprovedOptionalPermissions(
-        promptRequest: WebExtensionPromptRequest.AfterInstallation.OptionalPermissions,
-    ) {
-        promptRequest.onConfirm(true)
-        consumePromptRequest()
-    }
-
-    private fun handleDeniedOptionalPermissions(
-        promptRequest: WebExtensionPromptRequest.AfterInstallation.OptionalPermissions,
-    ) {
-        promptRequest.onConfirm(false)
+    private fun handlePermissions(promptRequest: WebExtensionPromptRequest.AfterInstallation.Permissions, granted: Boolean) {
+        promptRequest.onConfirm(granted)
         consumePromptRequest()
     }
 ```

Comment (willdurand) [reply fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt:330]: loevely, thanks!

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt [hunk 2]
```diff
@@ -38,27 +38,33 @@ sealed class WebExtensionPromptRequest {
...
-        ) : AfterInstallation(extension)
 
-        /**
-         * Value type that represents a request for a (optional) permission prompt.
-         * @property extension The [WebExtension] that requested the dialog to be shown.
-         * @property permissions The optional permissions to list in the dialog.
-         * @property onConfirm A callback indicating whether the permissions were granted or not.
-         */
-        data class OptionalPermissions(
+        sealed class Permissions(
```
Comment (Amejia481) [inline android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt:42]: Nit: Let's add some comments for the new class :)

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt [hunk 3]
```diff
@@ -38,27 +38,33 @@ sealed class WebExtensionPromptRequest {
...
             override val extension: WebExtension,
-            val permissions: List<String>,
-            val onConfirm: (Boolean) -> Unit,
-        ) : AfterInstallation(extension)
+            open val onConfirm: (Boolean) -> Unit,
+        ) : AfterInstallation(extension) {
+            /**
+             * Value type that represents a request for a permission prompt.
```
Comment (Amejia481) [inline android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt:47]: Should we add required on the docs as well?

File: fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt [hunk 4]
```diff
@@ -235,36 +232,31 @@ class WebExtensionPromptFeature(
...
+            onPositiveButtonClicked = { handlePermissions(promptRequest, granted = true) },
+            onNegativeButtonClicked = { handlePermissions(promptRequest, granted = false) },
         )
         dialog.show(
             fragmentManager,
             PERMISSIONS_DIALOG_FRAGMENT_TAG,
         )
     }
 
+    @Suppress("NestedBlockDepth")
```
Comment (Amejia481) [inline fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt:244]: Nit: Maybe we don't need the suppression anymore as we are not adding more lines :)

**Pull request merged**